solver/llbsolver/solver.go

```diff
@@ -193,138 +197,189 @@ func (s *Solver) Solve(ctx context.Context, id string, sessionID string, req fro
 	}
 
 	var exporterResponse map[string]string
+	cacheExporters := exp.CacheExporters
 	if e := exp.Exporter; e != nil {
-		cached, err := result.ConvertResult(res, func(res solver.ResultProxy) (solver.CachedResult, error) {
-			return res.Result(ctx)
-		})
+		var inlineExporter *RemoteCacheExporter
+		cacheExporters, inlineExporter = splitExporters(cacheExporters)
+		exporterResponse, err = exportInlineCache(ctx, e, inlineExporter, j, res)
 		if err != nil {
 			return nil, err
 		}
+	}
 
-		inp, err := result.ConvertResult(cached, func(res solver.CachedResult) (cache.ImmutableRef, error) {
-			workerRef, ok := res.Sys().(*worker.WorkerRef)
-			if !ok {
-				return nil, errors.Errorf("invalid reference: %T", res.Sys())
-			}
-			return workerRef.ImmutableRef, nil
-		})
-		if err != nil {
-			return nil, err
+	cacheExporterResponse, err := runCacheExporters(ctx, cacheExporters, j, res)
+	if err != nil {
+		return nil, err
+	}
+
+	if exporterResponse == nil {
+		exporterResponse = make(map[string]string)
+	}
+
+	for k, v := range res.Metadata {
+		if strings.HasPrefix(k, "frontend.") {
+			exporterResponse[k] = string(v)
+		}
+		if strings.HasPrefix(k, exptypes.ExporterBuildInfo) {
+			exporterResponse[k] = base64.StdEncoding.EncodeToString(v)
 		}
-		if res.Refs != nil {
-			m := make(map[string]cache.ImmutableRef, len(res.Refs))
-			for k, res := range res.Refs {
-				if res == nil {
-					m[k] = nil
-				} else {
+	}
+	for k, v := range cacheExporterResponse {
+		if strings.HasPrefix(k, "cache.") {
+			exporterResponse[k] = v
+		}
+	}
+
+	return &client.SolveResponse{
+		ExporterResponse: exporterResponse,
+	}, nil
+}
+
+func runCacheExporters(ctx context.Context, exporters []RemoteCacheExporter, j *solver.Job, res *frontend.Result) (map[string]string, error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	g := session.NewGroup(j.SessionID)
+	var cacheExporterResponse map[string]string
+	respCh := make(chan map[string]string, len(exporters))
+	for _, exp := range exporters {
+		exp := exp
+		eg.Go(func() error {
+			return inBuilderContext(ctx, j, "exporting cache", j.SessionID+"-cache", func(ctx context.Context, _ session.Group) error {
```

Comment:
We should change `"exporting cache"` to also include the cache type here (probably shouldn't have any of the other parameters, though we possibly could), and make sure that the ID field (currently `j.SessionID + "-cache"`) is unique for each exporter, or we'll get confusing build log output.

Personally, I'd like to have a `.Name` method on the [remotecache.Exporter](https://github.com/moby/buildkit/blob/master/cache/remotecache/export.go#L26-L32) interface, similar to how we already have for [exporter.ExporterInstance](https://github.com/moby/buildkit/blob/master/exporter/exporter.go#L18).

Reply:
Agree - I'll add the Name API on the `remotecache.Exporter`.